turn_rasa_connector/turn.py

```diff
@@ -136,13 +139,62 @@ def from_credentials(cls, credentials: Dict[Text, Any]) -> InputChannel:
             cls.raise_missing_credentials_exception()
 
         return cls(
-            credentials.get("hmac_secret"), credentials["url"], credentials["token"]
+            credentials.get("hmac_secret"),
+            credentials["url"],
+            credentials["token"],
+            credentials.get("postgresql_url"),
         )
 
-    def __init__(self, hmac_secret: Optional[Text], url: Text, token: Text) -> None:
+    def __init__(
+        self,
+        hmac_secret: Optional[Text],
+        url: Text,
+        token: Text,
+        postgresql_url: Optional[Text],
+    ) -> None:
         self.hmac_secret = hmac_secret
         self.url = url
         self.token = token
+        self.postgresql_url = postgresql_url
+        self._postgresql_pool = None
+
+    async def get_postgresql_pool(self) -> Optional[asyncpg.pool.Pool]:
+        if self._postgresql_pool is None and self.postgresql_url is not None:
+            self._postgresql_pool = await asyncpg.create_pool(self.postgresql_url)
+        return self._postgresql_pool
+
+    async def message_processed(
+        self, sender_id: Optional[Text], message_id: Optional[Text]
+    ) -> bool:
+        """
+        Have we processed a message with this ID before
+        """
+        if not sender_id or not message_id:
+            return False
+
+        pool = await self.get_postgresql_pool()
+        if pool is None:
+            # If we don't have a postgresql config, don't deduplicate
+            return False
+        async with pool.acquire() as connection:
+            async with connection.transaction():
+                result = await connection.fetchval(
+                    """
+                    SELECT 1
+                    FROM events
+                    WHERE
+                        sender_id = $1 AND
+                        type_name = $2 AND
+                        data::json ->> 'message_id' = $3 AND
```

Comment:
Not relevant to this particular line, but this feels like we're fixing the problem downstream when the root cause is upstream. Any reason why Turn would think this is ok behaviour to send you the same message with the same message id multiple times?

Reply:
When we last investigated, it was very difficult on our side with Turn forms to be able to get the debugging information that we needed to fully investigate it, but the feedback from Turn's side was that retries should only be happening on HTTP response failure, and HTTP requests that take too long to respond, but we couldn't confirm if there was something else going wrong.

In general, this is a difficult issue to solve, and Turn has gone with the "at least once" delivery model , which can lead to duplicate deliveries, instead of the "at most once", which can lead to lost messages.  The "exactly once" delivery model is difficult, and I'm not sure that it's possible to implement over HTTP webhooks.

Usually the way you solve this is by using the "at least once" model, but make sure that the task is idempotent. We can do that here, because we have an idempotency key, the message ID. So by deduplicating on the message ID, we can get an "exactly once" processing, with the "at least once" delivery model.